File: bench-adapters/js/benches/hamming.bench.js
// Head-to-head: StringCheese (via wasm) vs. JS Hamming implementations.
//
// Contestants
// ===========
//
// * **StringCheese** — Hamming from the wasm component. The WIT
//   boundary returns `result<u32, string>`; the adapter throws
//   `HammingLengthMismatch` on unequal lengths. Every input in this
//   file is equal-length by construction, so the error path is never
//   hit during timing.
// * **natural.HammingDistance** — a general-purpose NLP toolkit's
//   implementation.
// * **hand-rolled reference** — `[...a].filter((c, i) => c !== b[i]).length`,
//   the one-liner most JS answers reach for. Included to give the
//   reader a "what does the naive one-liner cost" baseline; the
//   ecosystem's dedicated Hamming implementations are surprisingly
//   thin on the ground, so an in-file reference is more informative
//   than pulling in a second npm package.
//
// Notable ecosystem gap: `fastest-levenshtein`, `js-levenshtein`, and
// `string-similarity` **do not** ship a Hamming function. If the
// downstream reader is looking for a Hamming-only fast path in the
// pure-JS ecosystem, `natural` is essentially the only turnkey option;
// everyone else writes it inline. That gap is part of the answer to
// "should I use StringCheese-via-wasm for Hamming from JS".
//
// Representation caveat
// =====================
//
// Same as the Levenshtein bench: StringCheese consumes `Uint8Array`,
// natural consumes `string`, and the hand-rolled reference is fed the
// same string. On ASCII input this is semantically equivalent; the FFI
// cost is folded in on purpose.

// `natural` transitively pulls in `dotenvx`; see levenshtein.bench.js
// for the ESM-hoisting rationale behind the side-effect import.
import "./_env.js";

import natural from "natural";

import { StringCheese } from "../stringcheese_adapter.js";
import { LENGTHS, REGIMES, buildPairEqualLen, lenLabel } from "./_inputs.js";
import { makeBench, runAll } from "./_runner.js";

const { HammingDistance: NaturalHamming } = natural;

// Shared with the Rust and Python adapters — Hamming needs equal-length
// inputs on both sides, and using the same salts across adapters means
// the mismatch positions line up for cross-adapter debugging.
const SALTS = /** @type {[number, number, number]} */ ([0xc1, 0xc2, 0xc3]);

/**
 * Naive-baseline Hamming: iterate one side, tally mismatches. Fed
 * strings because that is what a StackOverflow-copy-paste would look
 * like — a `Uint8Array` version would win on speed but would not
 * reflect what the ecosystem actually writes when it needs Hamming.
 * @param {string} a
 * @param {string} b
 * @returns {number}
 */
function handRolledHamming(a, b) {
  if (a.length !== b.length) {
    throw new Error("Hamming: length mismatch");
  }
  let n = 0;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) n++;
  return n;
}

async function main() {
  const sc = new StringCheese();
  const decoder = new TextDecoder("ascii");

  /** @type {Map<string, {aB: Uint8Array, bB: Uint8Array, aS: string, bS: string}>} */
  const pairs = new Map();
  for (const n of LENGTHS) {
    for (const kind of REGIMES) {
      const [aB, bB] = buildPairEqualLen(n, kind, SALTS);
      pairs.set(`${kind}/${lenLabel(n)}`, {
        aB,
        bB,
        aS: decoder.decode(aB),
        bS: decoder.decode(bB),
      });
    }
  }

  const benches = [];
  for (const kind of REGIMES) {
    for (const n of LENGTHS) {
      const key = `${kind}/${lenLabel(n)}`;
      const { aB, bB, aS, bS } = pairs.get(key);
      const bench = makeBench(`hamming/${key}`);
      bench.add("stringcheese", () => sc.hammingDistance(aB, bB));
      bench.add("natural", () => NaturalHamming(aS, bS));
      bench.add("hand-rolled", () => handRolledHamming(aS, bS));
      benches.push(bench);
    }
  }

  await runAll(benches);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: bench-adapters/python/benches/hamming_bench.py
"""Head-to-head: StringCheese (via wasm) vs. Python's Hamming implementations.

# Contestants

* **StringCheese** — Hamming from the wasm component. The WIT boundary
  returns ``result<u32, string>``; the adapter raises
  :class:`HammingLengthMismatch` on unequal lengths. Every input in this
  file is equal-length by construction, so the error path is never hit
  during timing.
* **python-Levenshtein.hamming** — C extension over Python strings.
* **jellyfish.hamming_distance** — Rust extension over Python strings.

rapidfuzz **is not** included in this file: as of ``rapidfuzz 3.14``
Hamming is exposed but the calling convention is
``rapidfuzz.distance.Hamming.distance(s1, s2, *, pad=False)``; the
default ``pad=False`` matches strsim/jellyfish/python-Levenshtein
semantics and would slot in cleanly, but the ecosystem-standard-pair
for Hamming in Python is still ``Levenshtein.hamming`` +
``jellyfish.hamming_distance``. Adding rapidfuzz here is a followup;
the group scaffolding is fixed and additive.

# Representation caveat

Same as the Levenshtein bench: StringCheese consumes ``bytes``, the two
native libraries consume ``str``. On ASCII input this is
semantically equivalent; the FFI cost is folded in on purpose.
"""

from __future__ import annotations

import Levenshtein
import jellyfish
import pytest

from _inputs import LENGTHS, REGIMES, build_pair_equal_len

_SALTS = (0xC1, 0xC2, 0xC3)


@pytest.fixture(scope="module")
def pairs() -> dict[tuple[int, str], tuple[bytes, bytes]]:
    """Equal-length pairs — Hamming requires the two sides to match in length."""
    return {(n, k): build_pair_equal_len(n, k, _SALTS) for n in LENGTHS for k in REGIMES}


# --------------------------------------------------------------------------- #
# StringCheese (via wasm)
# --------------------------------------------------------------------------- #


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_stringcheese(benchmark, stringcheese, pairs, n, kind):
    benchmark.group = f"hamming/{kind}/len{n:04d}"
    a, b = pairs[(n, kind)]
    benchmark(stringcheese.hamming_distance, a, b)


# --------------------------------------------------------------------------- #
# python-Levenshtein.hamming
# --------------------------------------------------------------------------- #


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_python_levenshtein(benchmark, pairs, n, kind):
    benchmark.group = f"hamming/{kind}/len{n:04d}"
    a_b, b_b = pairs[(n, kind)]
    a, b = a_b.decode("ascii"), b_b.decode("ascii")
    benchmark(Levenshtein.hamming, a, b)


# --------------------------------------------------------------------------- #
# jellyfish.hamming_distance
# --------------------------------------------------------------------------- #


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_jellyfish(benchmark, pairs, n, kind):
    benchmark.group = f"hamming/{kind}/len{n:04d}"
    a_b, b_b = pairs[(n, kind)]
    a, b = a_b.decode("ascii"), b_b.decode("ascii")
    benchmark(jellyfish.hamming_distance, a, b)
